File: conformance/binding-specs/grpc-runners/go/stack.go
package main

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"os"
	"path/filepath"
	"strings"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/credentials/insecure"
	_ "google.golang.org/grpc/encoding/gzip"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/emptypb"
	"google.golang.org/protobuf/types/known/wrapperspb"
)

type stackContract interface{ openBindingsStack() }
type stackServer struct{}

func (*stackServer) openBindingsStack() {}

func unaryHandler(srv any, ctx context.Context, decode func(any) error, interceptor grpc.UnaryServerInterceptor) (any, error) {
	in := new(emptypb.Empty)
	if err := decode(in); err != nil {
		return nil, err
	}
	call := func(ctx context.Context, _ any) (any, error) {
		values := metadata.ValueFromIncomingContext(ctx, "x-openbindings")
		if len(values) != 1 || values[0] != "stack" {
			return nil, status.Error(codes.InvalidArgument, "metadata")
		}
		if compressed := metadata.ValueFromIncomingContext(ctx, "x-compress"); len(compressed) == 1 && compressed[0] == "gzip" {
			if err := grpc.SetSendCompressor(ctx, "gzip"); err != nil {
				return nil, err
			}
		}
		return wrapperspb.String("unary"), nil
	}
	if interceptor == nil {
		return call(ctx, in)
	}
	return interceptor(ctx, in, &grpc.UnaryServerInfo{Server: srv, FullMethod: "/openbindings.test.Stack/Unary"}, call)
}

func serverHandler(_ any, stream grpc.ServerStream) error {
	if err := stream.RecvMsg(new(emptypb.Empty)); err != nil {
		return err
	}
	for _, value := range []string{"server-0", "server-1"} {
		if err := stream.SendMsg(wrapperspb.String(value)); err != nil {
			return err
		}
	}
	return nil
}

func clientHandler(_ any, stream grpc.ServerStream) error {
	var values []string
	for {
		message := new(wrapperspb.StringValue)
		err := stream.RecvMsg(message)
		if err == io.EOF {
			break
		}
		if err != nil {
			return err
		}
		values = append(values, message.Value)
	}
	return stream.SendMsg(wrapperspb.String(strings.Join(values, "+")))
}

func bidiHandler(_ any, stream grpc.ServerStream) error {
	for {
		message := new(wrapperspb.StringValue)
		err := stream.RecvMsg(message)
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err
		}
		if err := stream.SendMsg(wrapperspb.String("echo:" + message.Value)); err != nil {
			return err
		}
	}
}

var stackDescription = grpc.ServiceDesc{
	ServiceName: "openbindings.test.Stack",
	HandlerType: (*stackContract)(nil),
	Methods:     []grpc.MethodDesc{{MethodName: "Unary", Handler: unaryHandler}},
	Streams: []grpc.StreamDesc{
		{StreamName: "Server", Handler: serverHandler, ServerStreams: true},
		{StreamName: "Client", Handler: clientHandler, ClientStreams: true},
		{StreamName: "Bidi", Handler: bidiHandler, ServerStreams: true, ClientStreams: true},
	},
}

func stackFail(format string, values ...any) {
	fmt.Fprintf(os.Stderr, format+"\n", values...)
	os.Exit(1)
}

func main() {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		stackFail("listen: %v", err)
	}
	server := grpc.NewServer()
	server.RegisterService(&stackDescription, &stackServer{})
	go server.Serve(listener)
	defer server.Stop()

	connection, err := grpc.NewClient(listener.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		stackFail("client: %v", err)
	}
	defer connection.Close()
	ctx, cancel := context.WithTimeout(metadata.NewOutgoingContext(context.Background(), metadata.Pairs("x-openbindings", "stack")), 5*time.Second)
	defer cancel()

	unary := new(wrapperspb.StringValue)
	if err := connection.Invoke(ctx, "/openbindings.test.Stack/Unary", new(emptypb.Empty), unary); err != nil || unary.Value != "unary" {
		stackFail("unary: %v %q", err, unary.Value)
	}
	gzipCtx := metadata.AppendToOutgoingContext(ctx, "x-compress", "gzip")
	gzipUnary := new(wrapperspb.StringValue)
	if err := connection.Invoke(gzipCtx, "/openbindings.test.Stack/Unary", new(emptypb.Empty), gzipUnary, grpc.UseCompressor("gzip")); err != nil || gzipUnary.Value != "unary" {
		stackFail("gzip unary: %v %q", err, gzipUnary.Value)
	}

	serverStream, err := connection.NewStream(ctx, &grpc.StreamDesc{ServerStreams: true}, "/openbindings.test.Stack/Server")
	if err != nil {
		stackFail("server stream: %v", err)
	}
	if err := serverStream.SendMsg(new(emptypb.Empty)); err != nil {
		stackFail("server send: %v", err)
	}
	if err := serverStream.CloseSend(); err != nil {
		stackFail("server close: %v", err)
	}
	var serverValues []string
	for {
		message := new(wrapperspb.StringValue)
		err := serverStream.RecvMsg(message)
		if err == io.EOF {
			break
		}
		if err != nil {
			stackFail("server receive: %v", err)
		}
		serverValues = append(serverValues, message.Value)
	}
	if strings.Join(serverValues, ",") != "server-0,server-1" {
		stackFail("server values: %v", serverValues)
	}

	clientStream, err := connection.NewStream(ctx, &grpc.StreamDesc{ClientStreams: true}, "/openbindings.test.Stack/Client")
	if err != nil {
		stackFail("client stream: %v", err)
	}
	for _, value := range []string{"a", "b"} {
		if err := clientStream.SendMsg(wrapperspb.String(value)); err != nil {
			stackFail("client send: %v", err)
		}
	}
	if err := clientStream.CloseSend(); err != nil {
		stackFail("client close: %v", err)
	}
	clientReply := new(wrapperspb.StringValue)
	if err := clientStream.RecvMsg(clientReply); err != nil || clientReply.Value != "a+b" {
		stackFail("client reply: %v %q", err, clientReply.Value)
	}
	if err := clientStream.RecvMsg(new(wrapperspb.StringValue)); err != io.EOF {
		stackFail("client terminal: %v", err)
	}

	bidi, err := connection.NewStream(ctx, &grpc.StreamDesc{ServerStreams: true, ClientStreams: true}, "/openbindings.test.Stack/Bidi")
	if err != nil {
		stackFail("bidi stream: %v", err)
	}
	for _, value := range []string{"x", "y"} {
		if err := bidi.SendMsg(wrapperspb.String(value)); err != nil {
			stackFail("bidi send: %v", err)
		}
		reply := new(wrapperspb.StringValue)
		if err := bidi.RecvMsg(reply); err != nil || reply.Value != "echo:"+value {
			stackFail("bidi receive: %v %q", err, reply.Value)
		}
	}
	if err := bidi.CloseSend(); err != nil {
		stackFail("bidi close: %v", err)
	}
	if err := bidi.RecvMsg(new(wrapperspb.StringValue)); err != io.EOF {
		stackFail("bidi terminal: %v", err)
	}

	missing := connection.Invoke(ctx, "/openbindings.test.Stack/Missing", new(emptypb.Empty), new(emptypb.Empty))
	if status.Code(missing) != codes.Unimplemented {
		stackFail("missing status: %v", missing)
	}

	fixtureRoot := filepath.Join("..", "..", "grpc-fixtures", "tls")
	readFixture := func(name string) []byte {
		value, err := os.ReadFile(filepath.Join(fixtureRoot, name))
		if err != nil {
			stackFail("TLS fixture %s: %v", name, err)
		}
		return value
	}
	caBytes := readFixture("ca.pem")
	serverCertificate, err := tls.X509KeyPair(readFixture("server.pem"), readFixture("server-key.pem"))
	if err != nil {
		stackFail("server identity: %v", err)
	}
	clientCertificate, err := tls.X509KeyPair(readFixture("client.pem"), readFixture("client-key.pem"))
	if err != nil {
		stackFail("client identity: %v", err)
	}
	pool := x509.NewCertPool()
	if !pool.AppendCertsFromPEM(caBytes) {
		stackFail("test CA")
	}
	tlsListener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		stackFail("TLS listen: %v", err)
	}
	tlsServer := grpc.NewServer(grpc.Creds(credentials.NewTLS(&tls.Config{Certificates: []tls.Certificate{serverCertificate}, ClientAuth: tls.RequireAndVerifyClientCert, ClientCAs: pool, MinVersion: tls.VersionTLS13, MaxVersion: tls.VersionTLS13, NextProtos: []string{"h2"}})))
	tlsServer.RegisterService(&stackDescription, &stackServer{})
	go tlsServer.Serve(tlsListener)
	defer tlsServer.Stop()
	secureCredentials := credentials.NewTLS(&tls.Config{RootCAs: pool, ServerName: "api.example.com", Certificates: []tls.Certificate{clientCertificate}, MinVersion: tls.VersionTLS13, MaxVersion: tls.VersionTLS13, NextProtos: []string{"h2"}})
	secureConnection, err := grpc.NewClient(tlsListener.Addr().String(), grpc.WithTransportCredentials(secureCredentials))
	if err != nil {
		stackFail("TLS client: %v", err)
	}
	defer secureConnection.Close()
	secureReply := new(wrapperspb.StringValue)
	if err := secureConnection.Invoke(ctx, "/openbindings.test.Stack/Unary", new(emptypb.Empty), secureReply); err != nil || secureReply.Value != "unary" {
		stackFail("TLS unary: %v %q", err, secureReply.Value)
	}
	for label, config := range map[string]*tls.Config{
		"wrong-name":     {RootCAs: pool, ServerName: "wrong.example.com", Certificates: []tls.Certificate{clientCertificate}, MinVersion: tls.VersionTLS13, MaxVersion: tls.VersionTLS13, NextProtos: []string{"h2"}},
		"missing-client": {RootCAs: pool, ServerName: "api.example.com", MinVersion: tls.VersionTLS13, MaxVersion: tls.VersionTLS13, NextProtos: []string{"h2"}},
	} {
		candidate, err := grpc.NewClient(tlsListener.Addr().String(), grpc.WithTransportCredentials(credentials.NewTLS(config)))
		if err != nil {
			continue
		}
		candidateCtx, candidateCancel := context.WithTimeout(context.Background(), time.Second)
		failure := candidate.Invoke(candidateCtx, "/openbindings.test.Stack/Unary", new(emptypb.Empty), new(wrapperspb.StringValue))
		candidateCancel()
		candidate.Close()
		if failure == nil {
			stackFail("TLS %s unexpectedly succeeded", label)
		}
	}

	result := map[string]any{"format": "openbindings.grpc-real-stack-result@1", "runtime": "grpc-go", "unary": 1, "gzipUnary": 1, "serverOutputs": len(serverValues), "clientInputs": 2, "bidiInputs": 2, "bidiOutputs": 2, "status": "UNIMPLEMENTED", "tlsUnary": 1, "tlsFailures": 2}
	encoded, _ := json.Marshal(result)
	fmt.Println(string(encoded))
}

File: conformance/binding-specs/grpc-runners/typescript/stack.mjs
#!/usr/bin/env node

import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import grpc from "@grpc/grpc-js";
import protoLoader from "@grpc/proto-loader";

const here = dirname(fileURLToPath(import.meta.url));
const definition = protoLoader.loadSync(join(here, "stack.proto"), {
  defaults: false,
  enums: String,
  longs: String,
  oneofs: true
});
const loaded = grpc.loadPackageDefinition(definition);
const Stack = loaded.openbindings.test.Stack;

const implementation = {
  Unary(call, callback) {
    assert.deepEqual(call.metadata.get("x-openbindings"), ["stack"]);
    callback(null, { value: "unary" });
  },
  Server(call) {
    call.write({ value: "server-0" });
    call.write({ value: "server-1" });
    call.end();
  },
  Client(call, callback) {
    const values = [];
    call.on("data", (message) => values.push(message.value));
    call.on("end", () => callback(null, { value: values.join("+") }));
  },
  Bidi(call) {
    call.on("data", (message) => call.write({ value: `echo:${message.value}` }));
    call.on("end", () => call.end());
  }
};

const server = new grpc.Server({"grpc.default_compression_algorithm":grpc.compressionAlgorithms.gzip});
server.addService(Stack.service, implementation);
const address = await new Promise((resolve, reject) => server.bindAsync("127.0.0.1:0", grpc.ServerCredentials.createInsecure(), (error, port) => error ? reject(error) : resolve(`127.0.0.1:${port}`)));
const client = new Stack(address, grpc.credentials.createInsecure());
const gzipClient = new Stack(address,grpc.credentials.createInsecure(),{"grpc.default_compression_algorithm":grpc.compressionAlgorithms.gzip});
const metadata = new grpc.Metadata();
metadata.set("x-openbindings", "stack");
let tlsServer;

function unary(method, request, metadataValue = undefined) {
  return new Promise((resolve, reject) => client[method](request, metadataValue, (error, value) => error ? reject(error) : resolve(value)));
}
function collect(call) {
  return new Promise((resolve, reject) => { const values = []; call.on("data", (value) => values.push(value)); call.on("error", reject); call.on("end", () => resolve(values)); });
}

try {
  assert.equal((await unary("Unary", {}, metadata)).value, "unary");
  const gzipMetadata=new grpc.Metadata();gzipMetadata.set("x-openbindings","stack");gzipMetadata.set("x-compress","gzip");
  assert.equal((await new Promise((resolveValue,reject)=>gzipClient.Unary({},gzipMetadata,(error,value)=>error?reject(error):resolveValue(value)))).value,"unary");
  const serverValues = await collect(client.Server({}));
  assert.deepEqual(serverValues.map((value) => value.value), ["server-0", "server-1"]);

  const clientReply = new Promise((resolve, reject) => {
    const call = client.Client((error, value) => error ? reject(error) : resolve(value));
    call.write({ value: "a" }); call.write({ value: "b" }); call.end();
  });
  assert.equal((await clientReply).value, "a+b");

  const bidi = client.Bidi();
  const bidiValuesPromise = collect(bidi);
  bidi.write({ value: "x" }); bidi.write({ value: "y" }); bidi.end();
  const bidiValues = await bidiValuesPromise;
  assert.deepEqual(bidiValues.map((value) => value.value), ["echo:x", "echo:y"]);

  let missingStatus;
  try { await new Promise((resolve, reject) => client.makeUnaryRequest("/openbindings.test.Stack/Missing", (value) => value, (value) => value, Buffer.alloc(0), (error, value) => error ? reject(error) : resolve(value))); }
  catch (error) { missingStatus = error.code; }
  assert.equal(missingStatus, grpc.status.UNIMPLEMENTED);

  const fixtureRoot = resolve(here, "..", "..", "grpc-fixtures", "tls");
  const ca = readFileSync(join(fixtureRoot,"ca.pem"));
  const serverKey = readFileSync(join(fixtureRoot,"server-key.pem"));
  const serverCertificate = readFileSync(join(fixtureRoot,"server.pem"));
  const clientKey = readFileSync(join(fixtureRoot,"client-key.pem"));
  const clientCertificate = readFileSync(join(fixtureRoot,"client.pem"));
  tlsServer = new grpc.Server(); tlsServer.addService(Stack.service,implementation);
  const tlsAddress = await new Promise((resolveAddress,reject) => tlsServer.bindAsync("127.0.0.1:0",grpc.ServerCredentials.createSsl(ca,[{private_key:serverKey,cert_chain:serverCertificate}],true),(error,port)=>error?reject(error):resolveAddress(`127.0.0.1:${port}`)));
  const secureClient = new Stack(tlsAddress,grpc.credentials.createSsl(ca,clientKey,clientCertificate),{"grpc.ssl_target_name_override":"api.example.com","grpc.default_authority":"api.example.com"});
  assert.equal((await new Promise((resolveValue,reject)=>secureClient.Unary({},metadata,(error,value)=>error?reject(error):resolveValue(value)))).value,"unary");
  secureClient.close();
  const expectTlsFailure = async (label,credentials,options={}) => {
    const candidate = new Stack(tlsAddress,credentials,options);
    try { await new Promise((resolveValue,reject)=>candidate.Unary({},metadata,{deadline:Date.now()+1000},(error,value)=>error?reject(error):resolveValue(value))); throw new Error(`${label} unexpectedly succeeded`); }
    catch (error) { if (error.message?.includes("unexpectedly succeeded")) throw error; }
    finally { candidate.close(); }
  };
  await expectTlsFailure("wrong-name",grpc.credentials.createSsl(ca,clientKey,clientCertificate),{"grpc.ssl_target_name_override":"wrong.example.com","grpc.default_authority":"wrong.example.com"});
  await expectTlsFailure("missing-client",grpc.credentials.createSsl(ca),{"grpc.ssl_target_name_override":"api.example.com","grpc.default_authority":"api.example.com"});

  process.stdout.write(`${JSON.stringify({format:"openbindings.grpc-real-stack-result@1",runtime:"grpc-js",unary:1,gzipUnary:1,serverOutputs:serverValues.length,clientInputs:2,bidiInputs:2,bidiOutputs:bidiValues.length,status:"UNIMPLEMENTED",tlsUnary:1,tlsFailures:2})}\n`);
} finally {
  client.close();
  gzipClient.close();
  await new Promise((resolve) => server.tryShutdown(resolve));
  if (tlsServer) await new Promise((resolve) => tlsServer.tryShutdown(resolve));
}
